pkg/manifest/manifest_loader.go

```diff
@@ -0,0 +1,296 @@
+	var tokenType string
+
+	if token.Type == nil {
+		tokenType = "environment"
+	} else {
+		tokenType = *token.Type
+	}
+
+	switch tokenType {
+	case "environment":
+		return parseEnvironmentToken(token)
+	}
+
+	return nil, fmt.Errorf("unknown type `%s`", tokenType)
+}
+
+func parseEnvironmentToken(token tokenConfig) (Token, error) {
+	if val, found := token.Config["name"]; found {
+		return &EnvironmentVariableToken{
+			EnvironmentVariableName: toString(val),
+		}, nil
+	}
+
+	return nil, fmt.Errorf("missing `name` config")
+}
+
+func toString(v interface{}) string {
+	return fmt.Sprintf("%v", v)
+}
+
+func toProjectDefinitions(fs afero.Fs, workingDir string, definitions []project) (map[string]ProjectDefinition, []error) {
+	var errors []error
+	result := make(map[string]ProjectDefinition)
+
+	for _, project := range definitions {
+		parsed, projectErrors := parseProjectDefinition(fs, workingDir, project)
+
+		if projectErrors != nil {
+			errors = append(errors, projectErrors...)
+			continue
+		}
+
+		for _, project := range parsed {
+			if _, found := result[project.Name]; found {
+				errors = append(errors, fmt.Errorf("duplicated project `%s` defined in manifest", project.Name))
+				continue
+			}
+
+			result[project.Name] = project
+		}
+	}
+
+	if errors != nil {
+		return nil, errors
+	}
+
+	return result, nil
+}
+
+func parseProjectDefinition(fs afero.Fs, workingDir string, project project) ([]ProjectDefinition, []error) {
+	var projectType string
+
+	if project.Type == nil {
+		projectType = "simple"
+	} else {
+		projectType = *project.Type
+	}
+
+	switch projectType {
+	case "simple":
+		return parseSimpleProjectDefinition(project)
+	case "grouping":
+		return parseGroupingProjectDefinition(fs, workingDir, project)
+	default:
+		return nil, []error{fmt.Errorf("invalid project type `%s`", projectType)}
+	}
+}
+
+func parseSimpleProjectDefinition(project project) ([]ProjectDefinition, []error) {
+	if project.Path == "" {
+		return nil, []error{fmt.Errorf("project `%s` is missing a path", project.Name)}
+	}
+
+	return []ProjectDefinition{
+		{
+			Name: project.Name,
+			Path: project.Path,
+		},
+	}, nil
+}
+
+func parseGroupingProjectDefinition(fs afero.Fs, workingDir string, project project) ([]ProjectDefinition, []error) {
+	projectPath := filepath.FromSlash(project.Path)
+	fullProjectPath := filepath.Join(workingDir, projectPath)
+
+	files, err := afero.ReadDir(fs, fullProjectPath)
+
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	var result []ProjectDefinition
+
+	for _, file := range files {
+		if !file.IsDir() {
+			continue
+		}
+
+		result = append(result, ProjectDefinition{
+			Name: project.Name + "." + file.Name(),
+			Path: filepath.Join(projectPath, file.Name()),
+		})
+	}
+
+	if result == nil {
+		return nil, []error{fmt.Errorf("empty project path defined for `%s`: `%s`", project.Name, fullProjectPath)} // TODO should we really fail here?
+	}
+
+	return result, nil
+}
+
+func isYaml(filename string) bool {
```

Comment:
Seems like we have this same function on 3 different places. Maybe. moving to utils would be better solution?